ramips: add support for SNR-CPE-ME2-SFP

alealexpro100 opened the pull request:
SNR-CPE-ME2-SFP is a wireless WiFi 5 router manufactured by SNR/NAG company.

```Specification:
- SoC: MediaTek MT7621A
- CPU: 880MHz
- Flash: 16 MB (GD25Q127CSIG)
- RAM:  256 MB
- WLAN: 2.4 GHz, 5 GHz (MediaTek MT7615DN)
- Ethernet: 4x 10/100/1000 Mbps
- SFP cage (using RTL8211FS-CG)
- USB 3.0 port
- Power: 12 VDC, 2 A
```

- [x] Update using sysupgrade
- [x] GPIO LED and BUTTON
- [x] Wi-Fi
- [x] SFP
- [x] Ethernet
- [x] USB
- [x] Stability
- [x] Correct MAC
- [x] U-boot env

Builds link: https://disk.yandex.ru/d/2Q_AgZliMr3e-Q

csharper2005 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1937,6 +1937,18 @@ define Device/snr_snr-cpe-me2-lite
+  DEVICE_VENDOR := SNR
+  DEVICE_MODEL := SNR-CPE-ME2-SFP
+  UIMAGE_NAME := $$(DEVICE_MODEL)
+  DEVICE_PACKAGES := kmod-mt7603 kmod-mt7615e kmod-mt7663-firmware-ap \
```
Why do you need kmod-mt7603, kmod-mt7615e, kmod-mt7663-firmware-ap? kmod-mt7615-firmware should be enough.

csharper2005 commented on target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,215 @@
+	};
+};
+
+&pcie1 {
```
You can drop this block for 7615dn.

alealexpro100 commented:
Removed note `reading sfp eeprom is NOT working` because it is driver issue https://github.com/openwrt/openwrt/issues/12045.

csharper2005 commented on target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,200 @@
+
+&state_default {
+	gpio {
+		groups = "jtag", "wdt";
```
I would suggest checking it out. I think you need to add at least "uart" here. And maybe remove wdt.

alealexpro100 replied:
UART is not used, but available to be soldered.

csharper2005 replied:
This is not about uart on the board. You should specify all pin groups used as gpio here.
https://www.t-firefly.com/download/FireWRT/hardware/MT7621.pdf -> 2.4.1 GPIO Pin share scheme

alealexpro100 replied:
State_default is already written according to this datasheet and pins.

csharper2005 replied:
Thanks and sorry! I checked outdated version of your patch.

csharper2005 commented on target/linux/ramips/dts/mt7621_snr_cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,200 @@
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
```
Are the wireless MACs correct? I would suggest that you add MAC addresses table to the commit message. Here is an example - https://github.com/openwrt/openwrt/commit/f490295bf213c5e181d16cbe53c163c1c9983e45

alealexpro100 replied:
Added MACs table

csharper2005 replied:
Are these real MACs from the stock firmware?
```
+----------+------------------+-------------------+
| use      | address          | example           |
+----------+------------------+-------------------+
| Device   | label            | 00:00:00:11:00:00 |
| Ethernet | + 1              | 00:00:00:11:00:01 |
| 2g       | + 2              | 00:00:00:11:00:02 |
| 5g       | + 3              | 00:00:00:11:00:03 |
+----------+------------------+-------------------+
```

alealexpro100 replied:
Yes.

alealexpro100 replied:
Iteration is from stock image. These MACs are written for example, they are not real.

MaxS0niX commented on target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,200 @@
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		usb {
```
https://www.kernel.org/doc/html/latest/leds/leds-class.html
I propose to generalize

MaxS0niX commented on target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,200 @@
+&spi0 {
+	status = "okay";
+
+	flash@0 { // GD25Q127CSIG
```
Here and in the commit message
https://gist.github.com/alealexpro100/8d63f271dba592c1121c1f98f7ccb529#file-snr-log2-L109

alealexpro100 replied:
The chip soldered on board is named GD25Q127CSIG. Output shows other.

csharper2005 commented on target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,199 @@
+		vpn {
+			label = "green:vpn";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
```
https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml
Look at `function-enumerator`. We have more than one instance of the same function.

csharper2005 commented on target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,199 @@
+		led_sys: sys {
+			label = "green:sys";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
```
https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml
Look at function-enumerator. We have more than one instance of the same function.

DragonBluep commented on target/linux/ramips/dts/mt7621_snr_snr-cpe-me2-sfp.dts:
```diff
@@ -0,0 +1,199 @@
+		#size-cells = <0>;
+		usb-power {
+			gpio-export,name="usb-power";
+			gpio-export,output=<1>;
```
MT7621 USB power supply can be controlled by regulator. ref: https://github.com/openwrt/openwrt/commit/665c2154ef122d10dfffeca95538ebdff82653b5

alealexpro100 replied:
Thanks. Added regulator for usb to dts.

csharper2005 replied:
> Thanks. Added regulator for usb to dts.

Is it works? Can you switch off and switch on the power on a real USB device?

alealexpro100 replied:
Yes, it works. Using `echo '1-1' | tee /sys/bus/usb/drivers/usb/unbind` and `echo '1-1' | tee /sys/bus/usb/drivers/usb/bind`.

csharper2005 replied:
Good news! Thanks!

hauke commented:
Do you plan to add support for reading and writing the u-boot environment soon?

You can backport the patches needed to make the SFP slot fully work. You can add them to target/linux/generic/backport-5.15/

alealexpro100 commented:
> Do you plan to add support for reading and writing the u-boot environment soon?

No. I did not have success to access U-boot (UART did not work). 

> You can backport the patches needed to make the SFP slot fully work. You can add them to target/linux/generic/backport-5.15/

Unfortunately, kernel (even [upstream](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/drivers/net/phy/realtek.c)) does not have support for it. I tried to use these patches ([1](https://github.com/DasSecurity-HatLab/openwrt/blob/ec05735c7181ec12fac948cb229ab0937f367020/target/linux/ramips/patches-5.4/9902-rtl8211fs.patch) and [2](https://github.com/x-wrt/x-wrt/blob/a8384ead787f4eee8752813dfaf646c9cc8ed3d6/target/linux/ramips/patches-5.10/992-net-phy-realtek-RTL8211FS-support-for-fiber-mode.patch)), but it did not work (reading eeprom).

alealexpro100 committed:
ramips: add support for SNR-CPE-ME2-SFP

SNR-CPE-ME2-SFP is a wireless router with SFP cage manufactured by SNR/NAG company.

Specification:
- SoC: MediaTek MT7621A
- CPU: 880MHz
- Flash: 16 MB (GD25Q127CSIG)
- RAM:  256 MB
- WLAN: 2.4 GHz, 5 GHz (MediaTek MT7615DN)
- Ethernet: 4x 10/100/1000 Mbps
- SFP cage (using RTL8211FS-CG)
- USB 3.0 port
- Power: 12 VDC, 2 A

Flash instruction via TFTP:
1. Boot SNR-CPE-ME2 to recovery mode
  (press reset button and power on device, hold button for ~10 seconds)
2. Send firmware via TFTP client:
 TFTP Server address: 192.168.1.1
 TFTP Client address: 192.168.1.131
3. Wait ~120 seconds to complete flashing
4. Do sysupgrade using web-interface

MAC Addresses(stock)
--------------------
+----------+------------------+-------------------+
| use      | address          | example           |
+----------+------------------+-------------------+
| Device   | label            | 6A:C4:DD:xx:xx:28 |
| Ethernet | + 1              | 6A:C4:DD:xx:xx:29 |
| 2g       | + 2              | 6A:C4:DD:xx:xx:2A |
| 5g       | + 3              | 6A:C4:DD:xx:xx:2B |
+----------+------------------+-------------------+

Notes:
- Reading sfp eeprom is not supported [1] (driver issue). Stock image has the same situation.

References:
1. https://forum.openwrt.org/t/mt7621-and-reading-sfp-eeprom/152249

Signed-off-by: Aleksey Nasibulin <alealexpro100@ya.ru>
